pkgs/development/libraries/catboost/default.nix

```diff
@@ -0,0 +1,113 @@
+{ lib
+, config
+, stdenv
+, fetchFromGitHub
+, cmake
+, libiconv
+, llvmPackages
+, ninja
+, openssl
+, python3Packages
+, ragel
+, yasm
+, zlib
+, cudaSupport ? config.cudaSupport
+, cudaPackages ? {}
+, pythonSupport ? false
+}:
+
+stdenv.mkDerivation (finalAttrs: {
+  pname = "catboost";
+  version = "1.2.2";
+
+  src = fetchFromGitHub {
+    owner = "catboost";
+    repo = "catboost";
+    rev = "refs/tags/v${finalAttrs.version}";
+    hash = "sha256-A1zCIqPOW21dHKBQHRtS+/sstZ2o6F8k71lmJFGn0+g=";
+  };
+
+  patches = [
+    ./remove-conan.patch
+  ];
+
+  postPatch = ''
+    substituteInPlace cmake/common.cmake \
+      --replace  "\''${RAGEL_BIN}" "${ragel}/bin/ragel" \
+      --replace "\''${YASM_BIN}" "${yasm}/bin/yasm"
+
+    shopt -s globstar
+    for cmakelists in **/CMakeLists.*; do
+      sed -i "s/OpenSSL::OpenSSL/OpenSSL::SSL/g" $cmakelists
+      ${lib.optionalString (lib.versionOlder cudaPackages.cudaVersion "11.8") ''
+        sed -i 's/-gencode=arch=compute_89,code=sm_89//g' $cmakelists
+        sed -i 's/-gencode=arch=compute_90,code=sm_90//g' $cmakelists
+      ''}
+    done
+  '';
+
+  outputs = [ "out" "dev" ];
+
+  nativeBuildInputs = [
+    cmake
+    llvmPackages.bintools
+    ninja
+    (python3Packages.python.withPackages (ps: with ps; [ six ]))
+    ragel
+    yasm
+  ] ++ lib.optionals cudaSupport (with cudaPackages; [
+    cuda_nvcc
+  ]);
+
+  buildInputs = [
+    openssl
+    zlib
+  ] ++ lib.optionals stdenv.isDarwin [
+    libiconv
+  ] ++ lib.optionals cudaSupport (with cudaPackages; [
+    cuda_cudart
+    cuda_cccl
+    libcublas
+  ]);
+
+  env = {
+    CUDAHOSTCXX = lib.optionalString cudaSupport "${stdenv.cc}/bin/cc";
```

Comment:
We set this variable in `setup-cuda-hook.sh` (propagated by cuda_nvcc in natvieBuildInputs), so this should be a no-op. Also, if upstream uses `FindCUDAToolkit` and `enable_language(CUDA)`, then `CUDAHOSTCXX` is superseded by `CMAKE_CUDA_HOST_COMPILER` (which the hook should set as well)

Reply:
As reported in [the first comment](https://github.com/NixOS/nixpkgs/pull/235226#discussion_r1240595982), nvcc uses gcc instead of clang as the backendStdenv without it.
We will get the following error message if we do not set this environment variable.

```
catboost> [2892/4763] Building CUDA object library/cpp/cuda/wrappers/CMakeFiles/cpp-cuda-wrappers.dir/kernel.cu.o
catboost> FAILED: library/cpp/cuda/wrappers/CMakeFiles/cpp-cuda-wrappers.dir/kernel.cu.o 
catboost> /nix/store/mr3zd7dhyxdwr9y9gf6l74fah1rkjkji-cuda_nvcc-11.8.89/bin/nvcc -forward-unknown-to-host-compiler -ccbin=/nix/store/848ld56pzbywlxy1pv9xi5d8gsvzb6hl-gcc-wrapper-11.4.0/bin/c++ -DCATBOOST_OPENSOURCE=yes -I/build/source -I/build/source/build -I/build/source/contrib/libs/nvidia/thrust -I/build/source/contrib/libs/nvidia/cub -I/build/source/contrib/libs/linux-headers -I/build/source/contrib/libs/linux-headers/_nf -I/build/source/contrib/libs/cxxsupp/libcxx/include -I/build/source/contrib/libs/cxxsupp/libcxxrt/include -I/build/source/contrib/libs/clang14-rt/include -I/build/source/contrib/libs/zlib/include -I/build/source/contrib/libs/double-conversion -I/build/source/contrib/libs/libc_compat/include/readpassphrase -I/build/source/contrib/libs/libc_compat/reallocarray -I/build/source/contrib/libs/libc_compat/random -isystem /nix/store/mr3zd7dhyxdwr9y9gf6l74fah1rkjkji-cuda_nvcc-11.8.89/include -D_THREAD_SAFE -D_PTHREADS -D_REENTRANT -D_LARGEFILE_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__LONG_LONG_SUPPORTED -D_GNU_SOURCE -DLIBCXX_BUILDING_LIBCXXRT -D_FILE_OFFSET_BITS=64 -D_YNDX_LIBUNWIND_ENABLE_EXCEPTION_BACKTRACE -DSSE_ENABLED=1 -DSSE3_ENABLED=1 -DSSSE3_ENABLED=1 -DSSE41_ENABLED=1 -DSSE42_ENABLED=1 -DPOPCNT_ENABLED=1 -DCX16_ENABLED=1 --compiler-options -fexceptions,-fno-common,-fcolor-diagnostics,-fdebug-default-version=4,-ffunction-sections,-fdata-sections,-Wall,-Wextra,-Wno-parentheses,-Wno-implicit-const-int-float-conversion,-Wno-unknown-warning-option,-pipe,-fuse-init-array,-m64,-mpopcnt,-mcx16,-Woverloaded-virtual,-Wimport-preprocessor-directive-pedantic,-Wno-undefined-var-template,-Wno-return-std-move,-Wno-defaulted-function-deleted,-Wno-pessimizing-move,-Wno-deprecated-anon-enum-enum-conversion,-Wno-deprecated-enum-enum-conversion,-Wno-deprecated-enum-float-conversion,-Wno-ambiguous-reversed-operator,-Wno-deprecated-volatile --expt-extended-lambda --expt-relaxed-constexpr --compiler-options -std=c++14 -DTHRUST_IGNORE_CUB_VERSION_CHECK --threads 0 -O3 -DNDEBUG -std=c++14 -Xcompiler=-fPIC -nostdinc++ -DLIBCXX_BUILDING_LIBCXXRT -D_libunwind_ -gencode arch=compute_35,code=sm_35 -gencode arch=compute_50,code=compute_50 -gencode arch=compute_52,code=sm_52 -gencode arch=compute_60,code=compute_60 -gencode arch=compute_61,code=compute_61 -gencode arch=compute_61,code=sm_61 -gencode arch=compute_70,code=sm_70 -gencode arch=compute_70,code=compute_70 --ptxas-options=-v -lineinfo --use_fast_math -gencode=arch=compute_80,code=sm_80 -gencode=arch=compute_86,code=sm_86 -gencode=arch=compute_89,code=sm_89 -gencode=arch=compute_90,code=sm_90 -MD -MT library/cpp/cuda/wrappers/CMakeFiles/cpp-cuda-wrappers.dir/kernel.cu.o -MF library/cpp/cuda/wrappers/CMakeFiles/cpp-cuda-wrappers.dir/kernel.cu.o.d -x cu -c /build/source/library/cpp/cuda/wrappers/kernel.cu -o library/cpp/cuda/wrappers/CMakeFiles/cpp-cuda-wrappers.dir/kernel.cu.o
catboost> nvcc warning : incompatible redefinition for option 'compiler-bindir', the last value of this option was used
catboost> nvcc warning : The 'compute_35', 'compute_37', 'sm_35', and 'sm_37' architectures are deprecated, and may be removed in a future release (Use -Wno-deprecated-gpu-targets to suppress warning).
catboost> g++: error: unrecognized command-line option '-fcolor-diagnostics'
catboost> g++: error: unrecognized command-line option '-fdebug-default-version=4'
catboost> g++: error: unrecognized command-line option '-fuse-init-array'
catboost> g++: error: unrecognized command-line option '-Wimport-preprocessor-directive-pedantic'
catboost> g++: error: unrecognized command-line option '-fcolor-diagnostics'
catboost> g++: error: unrecognized command-line option '-fdebug-default-version=4'
catboost> g++: error: unrecognized command-line option '-fuse-init-array'
catboost> g++: error: unrecognized command-line option '-Wimport-preprocessor-directive-pedantic'
catboost> g++: error: unrecognized command-line option '-fcolor-diagnostics'
catboost> g++: error: unrecognized command-line option '-fdebug-default-version=4'
catboost> g++: error: unrecognized command-line option '-fuse-init-array'
catboost> g++: error: unrecognized command-line option '-Wimport-preprocessor-directive-pedantic'
catboost> g++: error: unrecognized command-line option '-fcolor-diagnostics'
catboost> g++: error: unrecognized command-line option '-fdebug-default-version=4'
catboost> g++: error: unrecognized command-line option '-fuse-init-array'
catboost> g++: error: unrecognized command-line option '-Wimport-preprocessor-directive-pedantic'
catboost> g++: error: unrecognized command-line option '-fcolor-diagnostics'
catboost> g++: error: unrecognized command-line option '-fdebug-default-version=4'
catboost> g++: error: unrecognized command-line option '-fuse-init-array'
catboost> g++: error: unrecognized command-line option '-Wimport-preprocessor-directive-pedantic'
catboost> g++: error: unrecognized command-line option '-fcolor-diagnostics'
catboost> g++: error: unrecognized command-line option '-fdebug-default-version=4'
catboost> g++: error: unrecognized command-line option '-fuse-init-array'
catboost> g++: error: unrecognized command-line option '-Wimport-preprocessor-directive-pedantic'
catboost> g++: error: unrecognized command-line option '-fcolor-diagnostics'
catboost> g++: error: unrecognized command-line option '-fdebug-default-version=4'
catboost> g++: error: unrecognized command-line option '-fuse-init-array'
catboost> g++: error: unrecognized command-line option '-Wimport-preprocessor-directive-pedantic'
catboost> g++: error: unrecognized command-line option '-fcolor-diagnostics'
catboost> g++: error: unrecognized command-line option '-fdebug-default-version=4'
catboost> g++: error: unrecognized command-line option '-fuse-init-array'
catboost> g++: error: unrecognized command-line option '-Wimport-preprocessor-directive-pedantic'
catboost> g++: error: unrecognized command-line option '-fcolor-diagnostics'
catboost> g++: error: unrecognized command-line option '-fdebug-default-version=4'
catboost> g++: error: unrecognized command-line option '-fuse-init-array'
catboost> g++: error: unrecognized command-line option '-Wimport-preprocessor-directive-pedantic'
catboost> g++: error: unrecognized command-line option '-fcolor-diagnostics'
catboost> g++: error: unrecognized command-line option '-fdebug-default-version=4'
catboost> g++: error: unrecognized command-line option '-fuse-init-array'
catboost> g++: error: unrecognized command-line option '-Wimport-preprocessor-directive-pedantic'
catboost> g++: error: unrecognized command-line option '-fcolor-diagnostics'
catboost> g++: error: unrecognized command-line option '-fdebug-default-version=4'
catboost> g++: error: unrecognized command-line option '-fuse-init-array'
catboost> g++: error: unrecognized command-line option '-Wimport-preprocessor-directive-pedantic'
catboost> fatal   : Could not open input file /build/tmpxft_0000501d_00000000-16_kernel.compute_35.cpp1.ii
```